### PR TITLE
fix #22926, raw string escaping rules that allow writing all strings

### LINKED ISSUE #22926: Parsing of backslashes in custom string literals seems inconsistent
```
julia> parse(readline(STDIN))
raw"\""

:(@raw_str "\"")

julia>

julia> parse(readline(STDIN))
raw"\\""
:($(Expr(:incomplete, "incomplete: invalid string syntax")))

julia> parse(readline(STDIN))
raw"\\\""
:(@raw_str "\\\\\"")

julia> dump(ans)
Expr
  head: Symbol macrocall
  args: Array{Any}((2,))
    1: Symbol @raw_str
    2: String "\\\\\""
  typ: Any
```
In particular there is a confusion about what backslash actually does. It certainly escapes `"`, but as we see in the second example, it also escapes itself. However, in the third example it suddenly doesn't appear to escape itself anymore. In particular, it does not seem to be possible to obtain a string literal that parses to `@raw_str "\\\""` at the moment.

The options I see are
1) Make `raw"\\""` parse to `@raw_str "\\\""` (rather than an error)
2) Make `raw"\\\""` parse to `@raw_str "\\\""`.
3) Decouple parsing from the content and always give what was actually written, i.e. make `raw"\""` parse to `@raw_str "\\\""` and `raw"\\\""` parse to `@raw_str "\\\\\\\""`
4) Disallow `"` in custom string literals entirely.
5) Do nothing.

FWIW, I think I prefer option 3.

Whatever the decision with respect to parsing, I think we should adjust raw_str to have the invariant that whenever `print(r"<x>")` is a valid expression for some sequence of characters `<x>`, the output of that expression is `<x>`. That's currently not the case in the above example:
```
julia> print(raw"\\\"")
\\"
julia> print(raw"\"")
"
```
This discrepancy was noted when raw_str was originally introduced and is documented, but it bothers me.

### PR DESCRIPTION
**Pull request by JeffBezanson**
I implemented @vtjnash 's proposal since that seemed to be the most popular.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -130,6 +130,10 @@ Language changes
   * Like `_`, variable names consisting only of underscores can be assigned,
     but accessing their values is deprecated ([#24221]).
 
+  * Raw string literal escaping rules have been changed to make it possible to write all strings.
+    The rule is that backslashes escape both quotes and other backslashes, but only when appearing
+    directly before a quote character ([#22926]).
```
Comment (StefanKarpinski) [inline NEWS.md]: "but only when a sequence of backslashes precedes a quote character. Thus, `2n` backslashes followed by a quote encodes `n` backslashes and the end of the literal while `2n+1` backslashes followed by a quote encodes `n` backslashes followed by a quote character."

File: base/strings/io.jl [hunk 2]
```diff
@@ -371,15 +371,23 @@ macro b_str(s); :(Vector{UInt8}($(unescape_string(s)))); end
     @raw_str -> String
 
 Create a raw string without interpolation and unescaping.
-The exception is that quotation marks still must be escaped.
+The exception is that quotation marks still must be escaped. Backslashes
+escape both question marks and other backslashes, but only when immediately
+preceding a quote character.
```
Comment (StefanKarpinski) [inline base/strings/io.jl]: Ditto.

**Review by StefanKarpinski — APPROVED** [review]

**Review by vtjnash — APPROVED** [review]

**Comment by vtjnash** [comment]
Looks like there's a couple other tests to fix though too:
https://github.com/JuliaLang/julia/blob/2028ae858a5d0367b8e8f93e57b16a843fd64e12/test/syntax.jl#L742-L745

File: base/strings/io.jl [hunk 3]
```diff
@@ -371,15 +371,25 @@ macro b_str(s); :(Vector{UInt8}($(unescape_string(s)))); end
     @raw_str -> String
 
 Create a raw string without interpolation and unescaping.
-The exception is that quotation marks still must be escaped.
+The exception is that quotation marks still must be escaped. Backslashes
+escape both question marks and other backslashes, but only when  a sequence
```
Comment (jmert) [inline base/strings/io.jl]: "question" -> "quotation" ?

Comment (JeffBezanson) [reply base/strings/io.jl]: Good catch.

**Review by Keno — APPROVED** [review]

**Pull request merged**